File: 078_subsets/Solution2/Solution.go
package main

import "fmt"

func subsets(nums []int) [][]int {
    arrLen := len(nums)
    var results [][]int

    for i := 0; i <= arrLen; i++ {
        used := make([]bool, arrLen)
        dfs(nums, &results, 0, arrLen, i, used)
    }

    return results
}

func dfs(nums []int, results *[][]int, start, arrLen, steps int, used []bool) {
    if steps == 0 {
        var tmpRes []int
        for i := 0; i < arrLen; i++ {
            if used[i] {
                tmpRes = append(tmpRes, nums[i])
            }
        }
        *results = append(*results, tmpRes)
        return
    }

    for i := start; i < arrLen; i++ {
        used[i] = true
        dfs(nums, results, i + 1, arrLen, steps - 1, used)
        used[i] = false
    }
}

func main() {
    nums := []int{1, 2, 3}
    results := subsets(nums)
    for _, result := range results {
        fmt.Println(result)
    }
}

File: 078_subsets/Solution2/Solution.py
class Solution:
    def subsets(self, nums: list) -> list:
        results = []
        arr_len = len(nums)

        def dfs(start: int, steps: int, used: list):
            if steps == 0:
                tmp_res = []
                for i in range(0, arr_len):
                    if used[i]: tmp_res.append(nums[i])
                results.append(tmp_res)
                return

            for i in range(start, arr_len):
                used[i] = True
                dfs(i + 1, steps - 1, used)
                used[i] = False

        for i in range(arr_len + 1):
            dfs(0, i, [False] * arr_len)

        return results


if __name__ == '__main__':
    nums = [1, 2, 3]
    results = Solution().subsets(nums)
    for result in results:
        print(result)

